mujoco_warp/_src/io.py

```diff
@@ -1303,6 +1304,547 @@ def reset_contact(
   )
 
 
+def set_const_fixed(m: types.Model, d: types.Data):
+  """Compute fixed quantities (independent of qpos0).
+
+  Computes:
+    - body_subtreemass: mass of body and all descendants (depends on body_mass)
+    - ngravcomp: count of bodies with gravity compensation (depends on body_gravcomp)
+
+  Args:
+    m: The model containing kinematic and dynamic information (device).
+    d: The data object containing the current state and output arrays (device).
+  """
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def init_subtreemass(
+    body_mass_in: wp.array2d(dtype=float),
+    body_subtreemass_out: wp.array2d(dtype=float),
+  ):
+    worldid, bodyid = wp.tid()
+    body_mass_id = worldid % body_mass_in.shape[0]
+    body_subtreemass_id = worldid % body_subtreemass_out.shape[0]
+    body_subtreemass_out[body_subtreemass_id, bodyid] = body_mass_in[body_mass_id, bodyid]
+
+  @nested_kernel(module="unique", enable_backward=False)
+  def accumulate_subtreemass(
+    bodyid: int,
+    body_parentid: wp.array(dtype=int),
+    body_subtreemass_io: wp.array2d(dtype=float),
+  ):
+    worldid = wp.tid()
+    body_subtreemass_id = worldid % body_subtreemass_io.shape[0]
+    parentid = body_parentid[bodyid]
+    body_subtreemass_io[body_subtreemass_id, parentid] += body_subtreemass_io[body_subtreemass_id, bodyid]
+
+  wp.launch(init_subtreemass, dim=(d.nworld, m.nbody), inputs=[m.body_mass], outputs=[m.body_subtreemass])
+  for bodyid in range(m.nbody - 1, 0, -1):
+    wp.launch(accumulate_subtreemass, dim=d.nworld, inputs=[bodyid, m.body_parentid], outputs=[m.body_subtreemass])
+
+  body_gravcomp_np = m.body_gravcomp.numpy()
+  m.ngravcomp = int((body_gravcomp_np > 0.0).any(axis=0).sum())
+
+
+def set_const_0(m: types.Model, d: types.Data):
+  """Compute quantities that depend on qpos0.
+
+  Computes:
+    - tendon_length0: tendon resting lengths
+    - dof_invweight0: inverse inertia for DOFs
+    - body_invweight0: inverse spatial inertia for bodies
+    - tendon_invweight0: inverse weight for tendons
+    - cam_pos0, cam_poscom0, cam_mat0: camera reference positions
```

Comment:
'cam_mat0' is not a position, so maybe 'camera reference positions' -> 'camera references'?